src/rust/engine/process_execution/src/bounded_tests.rs

```diff
@@ -158,6 +158,125 @@ async fn correct_semaphore_slot_ids_2() {
     assert_eq!(id7, 3);
 }
 
+#[tokio::test]
+async fn correct_semaphore_slot_ids_range() {
+    let sema = mk_semaphore(3);
+
+    let wait_duration = Duration::from_secs(5);
+    // Channels to signal when the processes should start and complete execution
+    let (send_start1, rcv_start1) = oneshot::channel();
+    let (send_start2, rcv_start2) = oneshot::channel();
+    let (send_start3, rcv_start3) = oneshot::channel();
+    let (send_start4, rcv_start4) = oneshot::channel();
+
+    let (send_complete1, rcv_complete1) = oneshot::channel();
+    let (send_complete2, rcv_complete2) = oneshot::channel();
+    let (send_complete3, rcv_complete3) = oneshot::channel();
+    let (send_complete4, rcv_complete4) = oneshot::channel();
+
+    // Process 1 (takes 2 slots)
+    let t1 = tokio::spawn(
+        sema.clone()
+            .with_acquired_range(2, 2, move |id| async move {
+                send_start1.send(()).unwrap();
+                rcv_complete1.await.unwrap();
+                id
+            }),
+    );
+
+    // Process 2 (takes 1 slot)
+    let t2 = tokio::spawn(
+        sema.clone()
+            .with_acquired_range(1, 1, move |id| async move {
+                send_start2.send(()).unwrap();
+                rcv_complete2.await.unwrap();
+                id
+            }),
+    );
+
+    // Wait for first two processes to start
+    if let Err(_) = timeout(wait_duration, rcv_start1).await {
+        panic!("Process 1 didn't start within timeout");
+    }
+    if let Err(_) = timeout(wait_duration, rcv_start2).await {
+        panic!("Process 2 didn't start within timeout");
+    }
+
+    assert_eq!(0, sema.available_permits());
+
+    // Process 3 (takes 2 slot, will need to wait for process 1 and 2 to complete)
+    let t3 = tokio::spawn(
+        sema.clone()
+            .with_acquired_range(2, 2, move |id| async move {
+                send_start3.send(()).unwrap();
+                rcv_complete3.await.unwrap();
+                id
+            }),
+    );
+
+    // Process 4 (takes 1 slot, should execute before process 3, but does not)
+    let t4 = tokio::spawn(
+        sema.clone()
+            .with_acquired_range(1, 1, move |id| async move {
+                send_start4.send(()).unwrap();
+                rcv_complete4.await.unwrap();
+                id
+            }),
+    );
+
+    // Release process 2 first, allowing process 4 to run
+    send_complete2.send(()).unwrap();
+    let id2 = t2.await.unwrap();
+
+    // Tokio appears to impliment strict FIFO fairness which has the downside that
```

Comment:
I _think_ this is fine from a correctness standpoint (though I need to review the preemption code again). I think ideally the scheduler would be a little smarter and have some deadlines attached to these priorities so it does't devolve into serial execution